### ISSUE TITLE
Use Python buffer protocol when converting Python objects to .NET arrays

### ISSUE THREAD

**Issue by lostmsu** [pythonnet/pythonnet#1838]
Copying data from NumPy array to C# is a common task. Python.NET implements a simple sequence-to-array copying mechanism, which uses iterators. It is extremely inefficient for large arrays of primitive types.

Now we have a better way - using `PyBuffer` and buffer protocol. We could detect, that Python object implements the buffer protocol and use it to copy when possible, and fallback to iteration otherwise.

**Comment by lostmsu** [comment]
Yep.

**Comment by bruceYangxianglin** [comment]
Copying data from  C# array to Python list or Numpy array is also a common task.  Is there a more efficient way?

**Comment by bjudkewitz** [comment]
Agreed, converting a python array to a C# array is a very common task, and it is currently extremely slow. It would be fantastic to make it faster. Is it currently possibly with Python (and Python.NET) only, without any C# programming?

**Comment by yahav6893** [comment]
Hi! I’m looking for some guidance on how to speed up copying from Python to C#. 
Would anyone be able to point me in the right way to do that?
Thanks ahead.

**Comment by chickenservice** [comment]
Hi, is someone currently working on this? If not I'd be happy to try. I've been using pythonnet at work successfully for years and would be glad to help. I'm familiar with the topic because I have already dealt with converting numpy arrays from/to C# arrays in an efficient manner. It'd be great to have this natively in pythonnet.

**Comment by filmor** [comment]
@chickenservice To my knowledge, no one is currently actively working on this, so give it a shot! An implementation would be greatly appreciated!

**Comment by chickenservice** [comment]
@filmor Perfect, is there anything besides the steps in CONTRIBUTING.md that I need to be aware of?
